accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+
+We could allow
+
+```dart
+extension MyInt on num {
+  int get double => this * s;
+  static int get random => 4;
+}
+```
+
+to introduce both a `double` instance getter on `num` instances and a `random` getter on `num` itself, usable as `var n = num.random;`
+
+However, while this is possible, not all `on` types are class or mixin types. It is not clear what it would mean to put static methods on `on` types of extension like:
+
+- `extension X on Iterable<int>`
+- `extension X<T extends Comparable<T>> on Iterable<T>`
+- `extension X on int Function(int)`
+- `extension X on FutureOr<int>`
+- `extension X on int?`
+
+For the first two, we could put the static members on the `Iterable` class, but since the extension does not apply to *all* iterables, it is not clear that this is correct.
+
+For `int Function(int)` and `FutureOr<int>`, it's unclear how to call such a static method at all. We can denote`int Function(int)` with a type alias, but putting static members on type aliases is a new concept.  We could put the static method on `Function`, but that's not particularly discoverable, and why not require that they are put on `Functon` explicitly. For `FutureOr`, we could allow static members on `FutureOr` (which is a denotable type), but again it seems spurious. For `int?`, we could put the method on `int`, but why not  just require that it's on `int`.
+
+The issue here is that the type patterns used by `on` are much more powerful than what is necessary to put static members on class types.
+
+It would probably be more readable to introduce a proper static extension declaration:
+
+```dart
+static extension Foo on int {  // or: extension Foo on static int 
+  int fromList(List l) => l.length;
+}
+
+...
+  print(int.fromList([1, 2])); // 2
+```
+
+where the `on` type must be something that can already have static methods. 
+
+The disadvantage is that if you want to introduce related functionality that is both static and instance methods on a class, then you need to write two extensions with different names.
+
+If we allow extension static declarations like these, we can also allow extension constructors.
+
+### Omitting Names For Local Extensions
+
+If an extension declaration is only used locally in a library, there is no need to worry about naming conflicts or overrides. In that case, then name identifier can be omitted.
+
+Example:
+
+```dart
+extension<T> on List<T> {
+  void quadruple() { ... }
+}
+```
+
+This is equivalent to giving the extension a fresh private name.
+
+The grammar then becomes:
+
+```ebnf
+<extension> ::= 
+  `extension' <identifier>? <typeParameters>? `on' <type> `?'? `{'
+     memberDeclaration*
+  `}'
+```
+
+This is a simple feature, but with very low impact. It only allows you to omit a single private name for an extension that is only used in a single library. Unless there is a documented demand for this feature, it doesn't seem worth the effort.
```

Comment:
For what it's worth, it is an annoying source of friction to come up with a name for an extension, and is probably even more annoying when it's only used locally. Anonymous extensions would help, though only for private ones.

I guess a simpler solution is for us to just establish a convention and tell people to follow it so they don't have to think too much about the name.